### PR TITLE
`ui_web` library

### PR DESCRIPTION
**Pull request by eyebrowsoffire**
This introduces a public `ui_web` library that contains web-specific flutter APIs. The first thing to incorporate into this is `UrlStrategy`, which will allow us to replace a lot of the magic JS interop handshaking between framework and engine into a formalized API.

### CODE REVIEW

File: lib/web_ui/lib/src/engine/navigation/history.dart
Comment (mdebbar) [inline lib/web_ui/lib/src/engine/navigation/history.dart:7]: How hard would it be to make this similar to the `package:ui/ui.dart` above? (i.e. `package:ui_web/ui_web.dart` or something like that). I guess it's just a matter of rewriting it in the sdk_rewriter?

File: lib/web_ui/lib/src/engine/navigation/history.dart
Comment (mdebbar) [inline lib/web_ui/lib/src/engine/navigation/history.dart:103]: I'm not sure I follow the reasoning for this change. Could you clarify?

Below, the override is still defined as `void Function(covariant DomPopStateEvent)`.

File: lib/web_ui/lib/src/engine/navigation/history.dart
Comment (mdebbar) [inline lib/web_ui/lib/src/engine/navigation/history.dart:186]: Can we keep it the name `event` or make it `Object? state`?

I have a feeling you were trying to change `DomPopStateEvent event ` ==> `Object? state` but forgot to finish the change, or reverted it partially.

File: lib/web_ui/lib/ui_web/src/ui_web.dart
Comment (mdebbar) [inline lib/web_ui/lib/ui_web/src/ui_web.dart:5]:
```suggestion
/// This library defines the web-specific additions that go along with dart:ui
```

File: web_sdk/BUILD.gn
Comment (mdebbar) [inline web_sdk/BUILD.gn:23]: 😂

File: web_sdk/BUILD.gn
Comment (mdebbar) [inline web_sdk/BUILD.gn:84]: It looks like this flag only affects the library name `$libraryName` vs `_$libraryName`. Could we just change library names (e.g. `library_name = "_engine"`) instead of introducing a new flag?

File: lib/web_ui/lib/src/engine/navigation/url_strategy.dart
Comment (mdebbar) [inline lib/web_ui/lib/src/engine/navigation/url_strategy.dart:12]: My understanding is this will go away after we migrate `flutter_web_plugins` to use the Dart API directly, correct?

**Review by mdebbar — COMMENTED** [review]
LGTM!

I don't have strong opinions on the library name, so I'll save you the bikeshedding from my end :)

File: lib/web_ui/lib/src/engine/navigation/history.dart
Comment (eyebrowsoffire) [reply lib/web_ui/lib/src/engine/navigation/history.dart:7]: This just gets erased by the rewriter, and in the top-level library file it says `import 'dart:ui_web' as ui_web;` so for the original source files we can import it any way we like as long as it is imported under the name `ui_web`. I can try to find a cleaner way to do this import.

File: lib/web_ui/lib/src/engine/navigation/url_strategy.dart
Comment (eyebrowsoffire) [reply lib/web_ui/lib/src/engine/navigation/url_strategy.dart:12]: Yep!

File: web_sdk/BUILD.gn
Comment (eyebrowsoffire) [reply web_sdk/BUILD.gn:23]: A man, a plan, a canal, panama

File: web_sdk/BUILD.gn
Comment (eyebrowsoffire) [reply web_sdk/BUILD.gn:84]: Not without doing a ton of changes across the codebase that actually change the original source files to all be underscore prefixed (and I think there is probably a reason it is this way, having to do with how package imports work vs library imports). The difference is that the original source files all treat it as `engine` and the transformed files treat it as `_engine`. It's not just a matter of changing the library name, because there is actually an "input library name" and "output library name". In the private case, that means adding an underscore, in the public case they are the same. So I can either have a `public` flag that tells us whether to add the underscore or not, or I can have an `inputLibraryName` and `outputLibraryName` or something like that that is specified by each transformation, but that seems too weird and generic and easy to mess up.

File: lib/web_ui/lib/src/engine/navigation/history.dart
Comment (eyebrowsoffire) [reply lib/web_ui/lib/src/engine/navigation/history.dart:186]: Hmmm... you're right, I kind of did a half-measure here. I think I can clean this up a bit.

File: lib/web_ui/lib/src/engine/navigation/history.dart
Comment (eyebrowsoffire) [reply lib/web_ui/lib/src/engine/navigation/history.dart:103]: For reasons that I don't really want to write a massive essay about right now (but would be happy to talk about more offline if you want) I don't want any of the surface area of this API to be JS interop objects. It gets complicated with `package:web` and some of the future stuff there.

File: lib/web_ui/lib/src/engine/navigation/history.dart
Comment (mdebbar) [reply lib/web_ui/lib/src/engine/navigation/history.dart:7]: On second look, I think `package:ui/src/ui_web.dart` should work already?

File: lib/web_ui/lib/src/engine/navigation/history.dart
Comment (eyebrowsoffire) [reply lib/web_ui/lib/src/engine/navigation/history.dart]: So what I ended up doing here is to basically unwrap the JS event in `JSUrlStrategy` before it actually gets to the darty `UrlStrategy` object, and so the pop state listener just takes the state itself. I think this is a bit better. For stuff on the framework side that uses actual JS interop to do the pop listener thing, they can just unwrap it there as well. The path through JSUrlStrategy remains unchanged, so I think this is non-breaking for the framework (but I'll find out as soon as I start testing against the framework side).

File: lib/web_ui/lib/src/engine/navigation/url_strategy.dart
Comment (mdebbar) [inline lib/web_ui/lib/src/engine/navigation/url_strategy.dart:112]: `HashUrlStrategy` (above) also needs a similar change. I'm surprised the analyzer isn't screaming at you since the method has a different signature now.

Comment (eyebrowsoffire) [reply lib/web_ui/lib/src/engine/navigation/url_strategy.dart:112]: Good call!

**Review by yjbanov — APPROVED** [review]
lgtm

File: lib/web_ui/lib/ui_web/src/ui_web.dart
Comment (eyebrowsoffire) [reply lib/web_ui/lib/ui_web/src/ui_web.dart:6]: I definitely don't want a nameless library though. This is a library that the framework (and maybe eventually users) will use by name.

Comment (kevmoo) [reply lib/web_ui/lib/ui_web/src/ui_web.dart:6]: See also: https://github.com/flutter/engine/pull/40669

Comment (eyebrowsoffire) [reply lib/web_ui/lib/ui_web/src/ui_web.dart:6]: The sdk_rewriter will not be able to rewrite this file properly without this statement.

**Comment by kevmoo** [comment]
That's okay. I don't think it's a huge deal

**Comment by eyebrowsoffire** [comment]
I'm going to wait until after the branch cut to merge this.

**Pull request merged**
